File: code/SungHo-Ck/week1/boj_2174.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int a, b, N, M;
int pass = 1;

struct robot {
    int x;
    int y;
    int num;
    char dir;
};

struct robot robots[100];
int map[101][101];

void wallB(int num) {
    printf("Robot %d crashes into the wall", num + 1);
	pass = 0;
}

void robotB(int num) {
	printf("Robot %d crashes into robot %d", num + 1, map[robots[num].x][robots[num].y] + 1);
	pass = 0;
}

void left(struct robot* robot) {
    switch (robot->dir) {
		case 'N':
			robot->dir = 'W';
			break;
		case 'W':
			robot->dir = 'S';
			break;
		case 'S':
			robot->dir = 'E';
			break;
		case 'E':
			robot->dir = 'N';
			break;
	}
}

void right(struct robot* robot) {
    switch (robot->dir) {
        case 'N':
            robot->dir = 'E';
            break;
        case 'W':
            robot->dir = 'N';
            break;
        case 'S':
            robot->dir = 'W';
            break;
        case 'E':
            robot->dir = 'S';
            break;
    }
}

void front(struct robot* r) {
    switch (r->dir) {
        case 'N':
            r->y++;
            if (r->y > b) {
                wallB(r->num);
                return;
            }
            else {
                if (map[r->x][r->y] != 0) {
                    robotB(r->num);
                    return;
                }
                else {
                    map[r->x][r->y] = r->num;
                    map[r->x][r->y - 1] = 0;
                }
            }
            break;
        case 'W':
            r->x--;
            if (r->x < 1) {
                wallB(r->num);
                return;
            }
            else {
                if (map[r->x][r->y] != 0) {
                    robotB(r->num);
                    return;
                }
                else {
                    map[r->x][r->y] = r->num;
                    map[r->x + 1][r->y] = 0;
                }
            }
            break;
        case 'S':
            r->y--;
            if (r->y < 1) {
                wallB(r->num);
                return;
            }
            else {
                if (map[r->x][r->y] != 0) {
                    robotB(r->num);
                    return;
                }
                else {
                    map[r->x][r->y] = r->num;
                    map[r->x][r->y + 1] = 0;
                }
            }
            break;
        case 'E':
            r->x++;
            if (r->x > a) {
                wallB(r->num);
                return;
            }
            else {
                if (map[r->x][r->y] != 0) {
                    robotB(r->num);
                    return;
                }
                else {
                    map[r->x][r->y] = r->num;
                    map[r->x - 1][r->y] = 0;
                }
            }
            break;
    }
}

int main(void) {
    scanf("%d %d", &a, &b);
    scanf("%d %d", &N, &M);
    for (int i = 0; i < N; i++) {
        int x, y;
        char dir;
        scanf("%d %d %c", &x, &y, &dir);
        robots[i].x = x;
        robots[i].y = y;
        robots[i].num = i;
        robots[i].dir = dir;
        map[x][y] = i;
    }
    for (int i = 0; i < M; i++) {
        if (!pass) break;
        int num, cnt;
        char order;
        scanf("%d %c %d", &num, &order, &cnt);
        switch (order) {
            case 'L':
                for (int i = 0; i < cnt; i++) {
                    if (!pass)
                        break;
                    left(&robots[num-1]);
                }
                break;
            case 'R':
                for (int i = 0; i < cnt; i++) {
                    if (!pass)
                        break;
                    right(&robots[num-1]);
                }
                break;
			case 'F':
                for (int i = 0; i < cnt; i++) {
                    if (!pass)
                        break;
                    front(&robots[num-1]);
                }
        }
    }

    if (pass) {
		printf("OK");
	}
    return 0;
}

File: code/jonghyeonjo99/week1/boj_2174.py
A, B = map(int,input().split())
N, M = map(int,input().split())

robots = []

# N W S E
dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]

for i in range(1,N+1):

  robot = list(input().split())

  #문제의 좌표 변환
  x = B - int(robot[1])
  y = int(robot[0]) - 1
  robot[0] = x
  robot[1] = y

  if robot[2] == 'N':
    robot[2] = 0
  elif robot[2] == 'W':
    robot[2] = 1
  elif robot[2] == 'S':
    robot[2] = 2
  else:           #'E'
    robot[2] = 3

  robots.append(robot)

for i in range(M):
  order = list(input().split())
  for j in range(int(order[2])):
    if order[1] == 'F':
      nx = robots[int(order[0])-1][0] + dx[robots[int(order[0])-1][2]]
      ny = robots[int(order[0])-1][1] + dy[robots[int(order[0])-1][2]]
      for k in robots:
        if k[0] == nx and k[1] == ny:
          print("Robot",order[0],"crashes into robot",robots.index(k)+1)
          exit()
      robots[int(order[0])-1][0] = nx
      robots[int(order[0])-1][1] = ny

      if nx < 0 or nx >= B or ny < 0 or ny >= A:
        print("Robot",order[0],"crashes into the wall")
        exit()
    elif order[1] == 'L':
      #robot 머리 숫자 + 1
      #if robot 머리 숫자 == 4 -> 0으로 바꿈
      robots[int(order[0])-1][2] += 1
      if robots[int(order[0])-1][2] == 4:
        robots[int(order[0])-1][2] = 0
    elif order[1] == 'R':
      #robot 머리 숫자 - 1
      #if robot 머리 숫자 == -1 -> 3으로 바꿈
      robots[int(order[0])-1][2] -= 1
      if robots[int(order[0])-1][2] == -1:
        robots[int(order[0])-1][2] = 3
print("OK")


